File: 0355-design-twitter/0355-design-twitter.cpp
class Twitter {
private:
    int timeStamp;
    unordered_map<int, unordered_set<int>> following;
    // userId -> {(time, tweetId)}
    unordered_map<int, vector<pair<int,int>>> tweets;
public:
    Twitter() {
        timeStamp = 0;
    }

    void postTweet(int userId, int tweetId) {
        tweets[userId].push_back({timeStamp++, tweetId});
    }
    vector<int> getNewsFeed(int userId) {
        priority_queue<vector<int>> pq;
        // user follows himself
        following[userId].insert(userId);
        for (int followee : following[userId]) {
            auto &userTweets = tweets[followee];
            if (!userTweets.empty()) {
                int idx = userTweets.size() - 1;
                pq.push({
                    userTweets[idx].first,   // timestamp
                    userTweets[idx].second,  // tweetId
                    followee,                // user
                    idx                      // index in vector
                });
            }
        }
        vector<int> feed;
        while (!pq.empty() && feed.size() < 10) {

            auto top = pq.top();
            pq.pop();

            int tweetId = top[1];
            int user = top[2];
            int idx = top[3];
            feed.push_back(tweetId);
            idx--;
            if (idx >= 0) {
                pq.push({
                    tweets[user][idx].first,
                    tweets[user][idx].second,
                    user,
                    idx
                });
            }
        }
        return feed;
    }
    void follow(int followerId, int followeeId) {
        following[followerId].insert(followeeId);
    }
    void unfollow(int followerId, int followeeId) {
        if (followerId == followeeId)
            return;
        following[followerId].erase(followeeId);
    }
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter* obj = new Twitter();
 * obj->postTweet(userId,tweetId);
 * vector<int> param_2 = obj->getNewsFeed(userId);
 * obj->follow(followerId,followeeId);
 * obj->unfollow(followerId,followeeId);
 */

File: 0355-design-twitter/0355-design-twitter.py
import heapq
class Twitter:
    def __init__(self):
        self.tweets = {}       # userId -> list of (time, tweetId)
        self.following = {}    # userId -> set of users they follow
        self.time = 0
    def postTweet(self, userId: int, tweetId: int) -> None:

        if userId not in self.tweets:
            self.tweets[userId] = []

        self.tweets[userId].append((self.time, tweetId))
        self.time += 1

    def getNewsFeed(self, userId: int) -> List[int]:

        heap = []
        # Include user's own tweets
        if userId in self.tweets:
            for time, tweetId in self.tweets[userId]:
                heapq.heappush(heap, (-time, tweetId))
        # Include tweets from followed users
        if userId in self.following:
            for followeeId in self.following[userId]:

                if followeeId in self.tweets:
                    for time, tweetId in self.tweets[followeeId]:
                        heapq.heappush(heap, (-time, tweetId))

        # Get 10 most recent tweets
        feed = []

        while heap and len(feed) < 10:
            time, tweetId = heapq.heappop(heap)
            feed.append(tweetId)

        return feed

    def follow(self, followerId: int, followeeId: int) -> None:

        if followerId not in self.following:
            self.following[followerId] = set()

        self.following[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:

        if followerId in self.following:
            self.following[followerId].discard(followeeId)
        


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)